crates/bitwarden-core/src/key_management/master_password.rs

```diff
@@ -3,39 +3,74 @@ use std::num::NonZeroU32;
 use bitwarden_api_api::models::{
     master_password_unlock_response_model::MasterPasswordUnlockResponseModel, KdfType,
 };
-use bitwarden_crypto::{EncString, Kdf};
+use bitwarden_crypto::{EncString, Kdf, MasterKey, SymmetricCryptoKey};
+use bitwarden_encoding::B64;
 use bitwarden_error::bitwarden_error;
 use serde::{Deserialize, Serialize};
+#[cfg(feature = "wasm")]
+use wasm_bindgen::prelude::*;
 
 use crate::{require, MissingFieldError};
 
 /// Error for master password related operations.
 #[allow(dead_code)]
 #[bitwarden_error(flat)]
 #[derive(Debug, thiserror::Error)]
-pub(crate) enum MasterPasswordError {
+pub enum MasterPasswordError {
     /// The wrapped encryption key could not be parsed because the encstring is malformed
     #[error("Wrapped encryption key is malformed")]
     EncryptionKeyMalformed,
     /// The KDF data could not be parsed, because it has an invalid value
     #[error("KDF is malformed")]
     KdfMalformed,
+    /// The KDF had an invalid configuration
+    #[error("Invalid KDF configuration")]
+    InvalidKdfConfiguration,
     /// The wrapped encryption key or salt fields are missing or KDF data is incomplete
     #[error(transparent)]
     MissingField(#[from] MissingFieldError),
+    /// Generic crypto error
+    #[error(transparent)]
+    Crypto(#[from] bitwarden_crypto::CryptoError),
 }
 
 /// Represents the data required to unlock with the master password.
-#[allow(dead_code)]
 #[derive(Serialize, Deserialize, Debug)]
 #[serde(rename_all = "camelCase", deny_unknown_fields)]
-pub(crate) struct MasterPasswordUnlockData {
+#[cfg_attr(feature = "uniffi", derive(uniffi::Record))]
+#[cfg_attr(
+    feature = "wasm",
+    derive(tsify::Tsify),
+    tsify(into_wasm_abi, from_wasm_abi)
+)]
+pub struct MasterPasswordUnlockData {
     /// The key derivation function used to derive the master key
-    kdf: Kdf,
+    pub kdf: Kdf,
     /// The master key wrapped user key
-    master_key_wrapped_user_key: EncString,
+    pub master_key_wrapped_user_key: EncString,
     /// The salt used in the KDF, typically the user's email
-    salt: String,
+    pub salt: String,
+}
+
+impl MasterPasswordUnlockData {
+    pub(crate) fn derive(
+        password: &str,
+        kdf: &Kdf,
+        salt: &str,
+        user_key: &SymmetricCryptoKey,
+    ) -> Result<Self, MasterPasswordError> {
+        let master_key = MasterKey::derive(password, salt, kdf)
+            .map_err(|_| MasterPasswordError::InvalidKdfConfiguration)?;
+        let master_key_wrapped_user_key = master_key
+            .encrypt_user_key(user_key)
+            .map_err(MasterPasswordError::Crypto)?;
+
+        Ok(Self {
+            kdf: kdf.clone(),
```

Comment:
(nit) Since it's a reference, we should use `to_owned()` instead of `clone()`
```suggestion
            kdf: kdf.to_owned(),
```